backend/cellguide/pipeline/computational_marker_genes/utils.py

```diff
@@ -0,0 +1,189 @@
+import os
+from typing import Tuple
+
+import numpy as np
+from numba import njit, prange
+from scipy import stats
+
+from backend.common.utils.rollup import are_cell_types_colinear
+from backend.wmg.data.utils import setup_retry_session
+
+
+@njit(parallel=True)
+def nanpercentile_2d(arr: np.ndarray, percentile: float, axis: int) -> np.ndarray:
+    """
+    Calculate the specified percentile of a 2D array along an axis, ignoring NaN values.
+
+    Arguments
+    ---------
+    arr - 2D array to calculate percentile of
+    percentile - percentile to calculate, as a number between 0 and 100
+    axis - axis along which to calculate percentile
+
+    Returns
+    -------
+    The specified percentile of the 2D array along the specified axis.
+    """
+    if axis == 0:
+        result = np.empty(arr.shape[1])
+        for i in prange(arr.shape[1]):
+            arr_column = arr[:, i]
+            result[i] = nanpercentile(arr_column, percentile)
+        return result
+    else:
+        result = np.empty(arr.shape[0])
+        for i in prange(arr.shape[0]):
+            arr_row = arr[i, :]
+            result[i] = nanpercentile(arr_row, percentile)
+        return result
+
+
+@njit
+def nanpercentile(arr: np.ndarray, percentile: float):
+    """
+    Calculate the specified percentile of an array, ignoring NaN values.
+
+    Arguments
+    ---------
+    arr - array to calculate percentile of
+    percentile - percentile to calculate, as a number between 0 and 100
+
+    Returns
+    -------
+    The specified percentile of the array.
+    """
+
+    arr_without_nan = arr[np.logical_not(np.isnan(arr))]
+    length = len(arr_without_nan)
+
+    if length == 0:
+        return np.nan
+
+    return np.percentile(arr_without_nan, percentile)
+
+
+def run_ttest(
+    *, sum1: np.ndarray, sumsq1: np.ndarray, n1: np.ndarray, sum2: np.ndarray, sumsq2: np.ndarray, n2: np.ndarray
+) -> Tuple[np.ndarray, np.ndarray]:
+    """
+    Run a t-test on two sets of data, element-wise.
+    Arrays "1" and "2" have to be broadcastable into each other.
+
+    Arguments
+    ---------
+    sum1 - Sum of the first set of data
+    sumsq1 - Sum of squares of the first set of data
+    n1 - Number of elements in the first set of data
+    sum2 - Sum of the second set of data
+    sumsq2 - Sum of squares of the second set of data
+    n2 - Number of elements in the second set of data
+
+    Returns
+    -------
+    pvals - The p-values of the t-test
+    effects - The effect sizes of the t-test
+    """
+
+    with np.errstate(divide="ignore", invalid="ignore"):
```

Comment:
do you think it would make sense to use a standard library for computing this instead of writing the logic ourselves?

https://www.geeksforgeeks.org/how-to-conduct-a-two-sample-t-test-in-python/#

i don't have a strong opinion on this particular use case since it's already written and appears to match this
![image](https://github.com/chanzuckerberg/single-cell-data-portal/assets/5653616/c13535b0-5046-4904-acc9-0deee115f90e)

but maybe something to keep in mind for the future?

Reply:
I don't think the standard library can do the t-test elementwise and with array broadcasting, but i could be wrong